### ISSUE TITLE
Building Razor views in release mode should minify HTML

### ISSUE THREAD

**Issue by RehanSaeed** [aspnet/Razor#423]
Minifying HTML is difficult in ASP.NET MVC. Comments and white space in .cshtml files could be removed when building in release mode.

<a href="http://www.deanhume.com/Home/BlogPost/a-simple-html-minifier-for-asp-net/2097">This</a> project allows the minification of your .cshtml files using a post build event. This should be built into the tooling and not a separate product.

**Comment by NTaylorMullen** [comment]
Hey @RehanSaeed are you suggesting to add Razor minification to the Visual Studio project templates?

**Comment by RehanSaeed** [comment]
In my head I imagine minification of cshtml being just a check-box in the project properties. It could then be turned off for Development and on for Staging and Production builds.

Adding this to the project as indicated in the blog is just one way of achieving this. I'm sure there are other ways which may be more integrated.

The benefits of minifying HTML are clearly shown by Mads Kristensen on his <a href="http://madskristensen.net/post/effects-of-gzipping-vs-minifying-html-files">blog article</a>. The best way of achieving minification is <strong>not</strong> during runtime as many frameworks do but during compile time where there is no performance hit to the application.

**Comment by NTaylorMullen** [comment]
Typically all minification occurs (in asp.net 5) via third party components that are configured to run post-build/restore etc. In my mind a .cshtml minifier would fall under the same bubble. With that mindset, if we were to add default .cshtml minification it would exist at the project template level.

**Comment by RehanSaeed** [comment]
It seems then that the solution employed by Dean Hume in his <a href="http://www.deanhume.com/Home/BlogPost/a-simple-html-minifier-for-asp-net/2097">blog article</a> is very close to what you are suggesting.

This would be a very cool feature if implemented. Most developers will not go to the added trouble of adding this feature themselves. It would probably save terabytes of bandwidth when you add up how many developers use ASP.NET MVC 6 and start using this feature because it comes right out of the box.

**Comment by NTaylorMullen** [comment]
@RehanSaeed thanks for the suggestion. Will discuss this with others to see if it has a place in our templates :smile:

**Comment by RehanSaeed** [comment]
Excellent. Thank You!

**Comment by glen-84** [comment]
I completely disagree with the suggested implementation.

Why not do this as part of [pre-compilation](https://github.com/aspnet/Mvc/issues/2551)? It can be an option in the `RazorViewEngineOptions` class.

I think that this issue should be re-opened.

/cc @GuardRex

**Comment by GuardRex** [comment]
@glen-84 :+1: I hate what I'm doing now, so that would be great.

However, @NTaylorMullen has there been any discussion on giving Razor directives line termination characters so that they wouldn't need to be on individual lines at the top of the markup and the Razor parser can grab them from minified markup? That's my main pain point with using Gulp to minify Razor views.

**Comment by NTaylorMullen** [comment]
@GuardRex no there hasn't. Not a horrible idea however, open an issue and we can discuss :smile:

**Comment by RehanSaeed** [comment]
I wrote a blog post about how much bandwidth this feature might have saved the internet [here](http://rehansaeed.com/minifying-html-for-asp-net-mvc/).

**Comment by ShamsulAmry** [comment]
May I know what's the status for this suggestion, is it denied? Suggestion by @glen-84 to have a precompiled minification would be really great. 

Scenario: I have templates which I bought containing lots of HTML and Javascript comments. I want to leave it in .cshtml because it helps during development, but I don't want / need those comments to appear in production.

On-the-fly minification can be done, but since the HTML and Javascript portions in a .cshtml is pretty much static most of the time, it would be great if those static portions can be minified during precompilation.

**Comment by GuardRex** [comment]
@ShamsulAmry Backlogged at https://github.com/aspnet/Razor/issues/444

**Comment by Sorrien** [comment]
Just wanted to throw in my support for built-in minification either at compile time or run-time.

**Comment by ffocsi** [comment]
Shouldn't this reopen?

I know for sure that knockout has a measurable performance increase (1 second boost in a real world project I have been working on) from html minification.

It's because when you have a
[ul foreach]
    [li with stuff][/li]
[/ul foreach]

The dom has [ul][text][li][text][ul]. Knockout has to process many more extra dom nodes to find bindings etc.

When minified it only sees [ul][li][ul].
(I found this out on a blog but can't find the source. But after profiling I discovered it was true and significant)

I'm using WebMarkupMin but this is done at runtime which impacts performance on a high load site. Would be nice if this could be done when the razor views get built. Would also be good with an option to do it in debug mode as well as release.

**Comment by guardrex** [comment]
@ffocsi [Razor Compile-time Minification for ASP.NET Core 2.0](http://www.guardrex.com/post/razor-minification.html)

**Comment by ffocsi** [comment]
Thanks.

Forgot to say we're still using asp.net 4 and mvc 5 for new projects.

I'm guessing we should upgrade.

**Comment by tompazourek** [comment]
For anybody interested in this, I built a tiny HTML minifier that can be used with MVC 5:

https://github.com/tompazourek/RazorHtmlMinifier.Mvc5

**Comment by lonix1** [comment]
@guardrex Your minifier is cool, and I'm using it (thanks!), but is there no way to do this as compile-time?

Your minifier works "during compile-time during runtime" if I understand correctly - because it's part of the pipeline. But if you're doing a bin deploy, surely this minification should be specified in the `.csproj` or something like that?

**Comment by guardrex** [comment]
@lonix1 I'm glad to hear it still works. lol ... I haven't looked at this subject for quite a while now.

That code replaces the Razor Template Engine. Since it should run with view precompilation, it should be working at compile time. There was a time when precomp only worked for framework-dependent deployments, but AFAIK they fixed that. https://github.com/aspnet/MvcPrecompilation/issues/102

It wouldn't be specified in the project file because that code interacts with the Razor build engine (as a replacement of the Razor Template Engine). It's a service registration to replace the OOB engine ...

```csharp
.ConfigureServices(services =>
{
    services.AddMvc();
    services.AddSingleton<RazorTemplateEngine, CustomRazorTemplateEngine>();
})
```

**Comment by Sora2455** [comment]
Just FYI, @guardrex 's minifier needs to be modified to work with the most recent version of ASP Core 2.
Details here: #https://github.com/aspnet/Razor/issues/2422

**Comment by K3n4** [comment]
Is this feature available on ASP Core 2.1 or you have denied it ?
